daml-lf/transaction/src/main/scala/com/daml/lf/Error.scala

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.lf
+package interpretation
+
+import com.daml.lf.data.Ref.{Location, Party, TypeConName}
+import com.daml.lf.transaction.{GlobalKey, NodeId}
+import com.daml.lf.language.Ast
+import com.daml.lf.value.Value
+import com.daml.lf.value.Value.ContractId
+
+/** Daml exceptions that should be reported to the user
+  */
+sealed abstract class Error extends Serializable with Product {
+  override def toString: String = s"$productPrefix(${productIterator.mkString("\n")}"
+}
+
+object Error {
+
+  /** Unhandled exceptions */
+  final case class UnhandledException(exceptionType: Ast.Type, value: Value[ContractId])
+      extends Error
+
+  /** User initiated error, via e.g. 'abort' or 'assert' */
+  final case class UserError(message: String) extends Error
+
+  final case class ContractNotFound(cid: Value.ContractId) extends Error
+
+  /** Template pre-condition (ensure) evaluated to false and the transaction
+    * was aborted.
+    */
+  final case class TemplatePreconditionViolated(
+      templateId: TypeConName,
+      optLocation: Option[Location],
+      arg: Value[ContractId],
+  ) extends Error
+
+  /** A fetch or an exercise on a transaction-local contract that has already
+    * been consumed.
+    */
+  final case class LocalContractNotActive(
```

Comment:
not your fault, this was bogus before but the error here isn’t right. This also applies to non-local contracts. What matters is that they have been archived in the current transaction (so the archive is "local" to this transaction) not where they have been created.